Rubberduck.VBEEditor/SafeComWrappers/SafeRedirectedEventedComWrapper.cs

```diff
@@ -13,73 +13,80 @@ public abstract class SafeRedirectedEventedComWrapper<TSource, TEventSource, TEv
         private const int NotAdvising = 0;
         private readonly object _lock = new object();
         private TEventSource _eventSource; // The event source
+        private TEventInterface _eventSink; // The event sink
         private IConnectionPoint _icp; // The connection point
-        private int _cookie = NotAdvising;     // The cookie for the connection
+        private int _cookie = NotAdvising; // The cookie for the connection
 
-        protected SafeRedirectedEventedComWrapper(TSource target, TEventSource eventSource, bool rewrapping = false) : base(target, rewrapping)
+        protected SafeRedirectedEventedComWrapper(TSource target, bool rewrapping = false) 
+            : base(target, rewrapping)
         {
-            _eventSource = eventSource;
         }
 
         protected override void Dispose(bool disposing)
         {
             DetachEvents();
-
-            Marshal.ReleaseComObject(_eventSource);
-            _eventSource = null;
-
             base.Dispose(disposing);
         }
 
-        public void AttachEvents()
+        protected void AttachEvents(IEventSource<TEventSource> eventSource, TEventInterface eventSink)
         {
+            Debug.Assert(eventSource != null, "Unable to attach events - eventSource is null");
+            Debug.Assert(eventSink != null, "Unable to attach events - eventSink is null");
+            if (eventSource == null || eventSink == null)
+            {
+                return;
+            }
+
+            // Only one event source and one event sink supported at any given time
+            DetachEvents();
```

Comment:
After thinking about it, I think I don't want that particular change here and on the original line 43-45 below. As discussed, the "last one wins" model would likely do more damage than the current "first one wins" model. The latecomer should be handed a no-op instead so that the first can be assured that it can clean itself up properly.

That said, the requirement to have only one source is separate issue, and we probably could stick in a `Debug.Assert` to warn of accidental re-attaching by a different source, which is most likely a bug that needs to be fixed because we want to minimize the contact between the .NET event consumers and the COM event sink.